components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,723 @@
+    // Used to smoothly update IO budgets
+    history_epoch_count: usize,
+    history_bytes: [usize; IOPriority::COUNT],
+}
+
+impl PriorityBasedIORateLimiterProtected {
+    fn new() -> Self {
+        PriorityBasedIORateLimiterProtected {
+            next_refill_time: Instant::now_coarse() + DEFAULT_REFILL_PERIOD,
+            pending_bytes: [0; IOPriority::COUNT],
+            history_epoch_count: 0,
+            history_bytes: [0; IOPriority::COUNT],
+        }
+    }
+}
+
+macro_rules! do_sleep {
+    ($duration:expr, sync) => {
+        std::thread::sleep($duration);
+    };
+    ($duration:expr, async) => {
+        tokio::time::delay_for($duration).await;
+    };
+    ($duration:expr, skewed_sync) => {
+        use rand::Rng;
+        let mut rng = rand::thread_rng();
+        let subtraction: bool = rng.gen();
+        let offset = std::cmp::min(Duration::from_millis(1), $duration / 100);
+        std::thread::sleep(if subtraction {
+            $duration - offset
+        } else {
+            $duration + offset
+        });
+    };
+}
+
+/// Actual implementation for requesting IOs from PriorityBasedIORateLimiter.
+/// An attempt will first be recorded. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will be queued and sleep until served.
+/// Macro is necessary to de-dup codes used both in async/sync functions.
+macro_rules! request_imp {
+    ($limiter:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        let cached_bytes_per_refill =
+            $limiter.bytes_per_epoch[priority_idx].load(Ordering::Relaxed);
+        // flow control is disabled when limit is zero
+        if cached_bytes_per_refill == 0 {
+            return $amount;
+        }
+        let amount = std::cmp::min($amount, cached_bytes_per_refill);
+        let bytes_through =
+            $limiter.bytes_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+        // we prefer not to partially return only a portion of requested bytes
+        if bytes_through <= cached_bytes_per_refill {
+            return amount;
+        }
+        let now = Instant::now_coarse();
+        let mut wait = Duration::from_millis(0);
+        // acquire a snapshot of pending bytes, which denotes a position of logical queue
+        // for this request to wait in.
+        let pending_snapshot = {
+            let mut locked = $limiter.protected.lock();
+            // already served partially by current epoch when consumption overflow bytes is
+            // smaller than requested amount
+            let remains = std::cmp::min(bytes_through - cached_bytes_per_refill, amount);
+            // when there is a recent refill, double check if bytes consumption has been reset
+            if now + DEFAULT_REFILL_PERIOD < locked.next_refill_time + Duration::from_millis(1)
+                && $limiter.bytes_through[priority_idx].fetch_add(remains, Ordering::Relaxed)
```

Comment:
The second condition `bytes_through + remains <= cached_bytes_per_epoch` alone is probably enough?

Reply:
This tries to optimize out a `fetch_add` for most cases.